### ISSUE TITLE
Template strings support in JS lexer

### ISSUE THREAD

**Issue by everyonesdesign** [python-babel/babel#329]
It seems that [ECMAScript2015's Tempate strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings) breaks Babel's Javascript parsing.

That's sad, because it's supported quite well in latest browsers.

It would be great if Babel supported this feature.

**Comment by akx** [comment]
Entirely possible to implement!

There are PRs in the pipeline to improve JSX and dotted name parsing (#310 and #298 respectively).

However, a question:

Are you planning to use template strings in conjunction with translation, like

```javascript
gettext(`Hi, ${name}!`)
// or to use the tag syntax,
i18n `Hi, ${name}!` 
```

or would you just like to see parser support for plain template strings?

**Comment by everyonesdesign** [comment]
Well, I planned to use this way:

```js
new ScrollUp({
	text: `▲<br><span> ${ _("back to top") } </span>`
});
```

But, actually, it's not the best idea to put markup directly to strings.
After checking out your examples I see that the issue is not so easy to implement due to multiple usage ways.
Anyway, it would be great if there was a best practice or a recommendation on its usage and we could stick to it.

**Comment by akx** [comment]
Ah, that (parsing the `${ }` expressions within templates) should also be possible to implement. I didn't think about that :)

But sure, this definitely looks like something to implement along with the other JavaScript extraction improvements! :)

**Comment by akx** [comment]
@everyonesdesign If you're around, could you do code review on #332? :)

**Comment by everyonesdesign** [comment]
@akx 
I'm not so good at Python, so I looked mostly at the tests changes in the MR. I don't see tests for the recursive template strings parse trick you mentioned in [your comment](https://github.com/python-babel/babel/issues/329#issuecomment-171595726). Is it planned to be implemented?

Anyway, new usecases of the translations in tests look really good, thanks for that great job.

**Comment by akx** [comment]
@everyonesdesign Yes, that support is not (yet) implemented; it could be done as further development just to get some template string support in. :)

**Comment by johbo** [comment]
Just ran into this problem as well, taking a look if I can find out how to improve the situation.

**Comment by johbo** [comment]
Hmm, found out that my issues were caused by not recognizing the jsx tags themselves, the change in #396 gets it good enough for my need.

While looking at this topic, I noticed that template strings can be nested:

```
`${condition ? `${varA}` : `${varB}`}`
```

Although this seems to produce odd results, it's probably not going to harm since the backticks come in pairs.

**Comment by petterm** [comment]
I found this issue while troubleshooting extracting strings like the one below. Any updates on this?
```js
const text = `${variable} - ${gettext('some translated text')}`;
```

**Comment by fdanielsen** [comment]
Any updates on this? PR #332 references this as something included in that merged improvement, but this issue is still open and in my testing extraction from template strings does not work.

Would be really nice to have support for this. We ended up with an untranslated string in production due to unawareness of this issue with extraction, and had to somewhat awkwardly restructure some code to ensure the gettext calls were outside template strings to make sure they got extracted.

**Comment by akx** [comment]
@fdanielsen Sorry to hear about the trouble! If I may inquire, how are you using template strings for translation at the moment?

**Comment by fdanielsen** [comment]
@akx Thanks for the quick response. In this case I was building up a string with some minor HTML formatting in addition to a translatable string:

```javascript
const description = `<em>${gettext('…')}</em>`;
```

I've now changed it to simply extract the `gettext` call to a separate variable declaration that then gets interpolated in the template string. But would be nice to avoid that. And since we use Prettier which tends to reformat these string concatenations it would be nice to avoid having to remember this "gotcha" with `gettext` calls.

**Comment by gitaarik** [comment]
I've had issues with this for a while, so I decided to hack around. I got the JS template expressions working. These:

    `${gettext('my text')}`

And it also works with nested expressions like:

    ${`some text ${gettext('my text')} more text`}

I created a fork. You can try it out:

    pip install https://github.com/gitaarik/babel/archive/javascript-improvements.zip

On this branch I also fixed another issue that I was facing. In a JS library that I use, there was regex with a `/` inside a character class `[a-z]`. The babel JS lexer thought this closed the regex, but JavaScript allows this as a normal match character within a character class, and does not close the regex. I updated the regex so that it will cover this exception.

Here you can see the diff of my fork:

https://github.com/python-babel/babel/compare/master...gitaarik:javascript-improvements?expand=1

It works for me, and I'm very happy about that. But I'm not sure what needs to be done to merge this in the babel master. I don't know how to run the tests and if there needs to be unit tests added and how to etc. Are there people that can help me with this? Maybe @akx ?
